packages/@markuplint/ml-core/src/ml-dom/document.ts

```diff
@@ -90,33 +89,40 @@ export default class MLDOMDocument<T extends RuleConfigValue, O = null> {
 	 * @param walker
 	 * @param skipWhenRuleIsDisabled
 	 */
-	async walkOn(
+	walkOn(
 		type: 'Element',
 		walker: Walker<T, O, MLDOMElement<T, O>>,
 		skipWhenRuleIsDisabled?: boolean,
-	): Promise<void>;
-	async walkOn(type: 'Text', walker: Walker<T, O, MLDOMText<T, O>>, skipWhenRuleIsDisabled?: boolean): Promise<void>;
-	async walkOn(
+	): void | Promise<void[]>;
+	walkOn(
+		type: 'Text',
+		walker: Walker<T, O, MLDOMText<T, O>>,
+		skipWhenRuleIsDisabled?: boolean,
+	): void | Promise<void[]>;
+	walkOn(
 		type: 'Comment',
 		walker: Walker<T, O, MLDOMComment<T, O>>,
 		skipWhenRuleIsDisabled?: boolean,
-	): Promise<void>;
-	async walkOn(
+	): void | Promise<void[]>;
+	walkOn(
 		type: 'ElementCloseTag',
 		walker: Walker<T, O, MLDOMElementCloseTag<T, O>>,
 		skipWhenRuleIsDisabled?: boolean,
-	): Promise<void>;
-	async walkOn(type: NodeType, walker: Walker<T, O, any>, skipWhenRuleIsDisabled: boolean = true): Promise<void> {
-		for (const node of this.nodeList) {
-			if (node instanceof MLDOMNode) {
-				if (skipWhenRuleIsDisabled && node.rule.disabled) {
-					continue;
-				}
-				if (node.is(type)) {
-					await walker(node);
+	): void | Promise<void[]>;
+	walkOn(type: NodeType, walker: Walker<T, O, any>, skipWhenRuleIsDisabled: boolean = true): void | Promise<void[]> {
+		// sync API won't rely on the result
+		return Promise.all(
+			this.nodeList.map(node => {
+				if (node instanceof MLDOMNode) {
+					if (skipWhenRuleIsDisabled && node.rule.disabled) {
+						return;
+					}
+					if (node.is(type)) {
+						return walker(node);
```

Comment:
OK, I'll refactor it to be sequentially.

Reply:
But by the way, the benefits of Promise is parcelling, of course maybe it can be done in a major version if we love parcelling.

Reply:
@JounQin 

Sorry, I don't know about "parceling".
What do you mean said "a major version"?

Reply:
Or sorry, I mean "parallel".

I mean if you think `parallel` will be better, it can be refactored like current in a major version. I'll change it to be sequentially for compatibility.

If you do not like it, we just keep it always sequentially.

Reply:
@JounQin 

I see. I know the parallel has a benefit.
But the sequenceable is has a benefit too.
The idea is nice.
So I want to improve that add an option that the plugin authors can choose sequential/parallel.
But I will not accept it becomes parallel only.
I WANT UNIVERSALITY. Ok?

Reply:
@YusukeHirao 

Done! It is always sequentially now.

Reply:
fixed in [`9a197a1` (#157)](https://github.com/markuplint/markuplint/pull/157/commits/9a197a14b07e4bebaa9d3c486348b2f4d01ec344)